076-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,447 @@
+
+> ![Cluster CA private key update flow](./images/076-cert-update-proposed.png)
+>
+> Fig 7: Proposed update CA certificate flow
+> The numbers in the diagram match the numbers in the full description of the steps.
+> The diagram omits the use of annotations and the EntityOperator, CruiseControl and KafkaExporter reconcilers, and the removal of trust of the old CA certificate and key.
+
+Notes:
+* The existing annotations strimzi.io/ca-cert-generation and strimzi.io/server-cert-hash are still used to keep track of the certificate being presented by the Kubernetes Pod and which generation of the Cluster CA signed that certificate.
+* The strimzi.io/ca-key-generation annotation is no longer needed, because it is being tracked in the trusted-certs Kubernetes Secret.
+
+### Future changes
+
+This proposal does not cover all changes needed to integrate with certificate management solutions.
+Instead it aims to make an improvement to the existing logic and act as a first step towards integrating some certificate management solution.
+The approach suggested aligns with the approach taken in the [closed proposal on CA abstraction][pr46].
+
+Other/Future proposals will be needed to cover:
+ - Refactoring the end-entity certificate issuance process, so that components request a certificate and asynchronously are issued a certificate.
+ - Refactoring the tracking of which certificates are signed by which CA certificates.
+   Currently, the notion of a "generation" is used, however if Strimzi is no longer managing the CA this concept will likely be replaced as this is not the usual way to identify a certificate and therefore will be tricky to use with a certificate management system.
+   In future it would be better to identify the CA certificate used to sign a particular end-entity certificate using fingerprints, since it is clear how a fingerprint relates to a certificate.
+   The state machine described in this proposal would not need any major changes to support this, it was just require the check for state transitions to compare fingerprints rather than generation ids.
+ - Refactoring the CA certificate management process to allow alternative implementations internally, for example using the cert-manager API.
+
+## Affected/not affected projects
+
+This affects the strimzi-kafka-operator project only.
+
+## Compatibility
+
+In this proposal cluster-ca-cert Kubernetes Secret is still utilised by users managing their own CA and by applications to trust the cluster.
+This allows us to change the internals for how the certificates are managed, without causing breaking changes for users.
+
+The strimzi.io/ca-key-generation annotation is no longer needed, because it is being tracked in the trusted-certs Kubernetes Secret.
+However, removing it from the component pods would result in unnecessary rolling updates when the operator is downgraded.
```

Comment:
Why would it result in unnecessary rolling updates?

Reply:
In step 1 of the operator downgrade the CA Reconciler would check the cluster-ca-key-generation on the Kafka and ZooKeeper pods and (incorrectly) determine that the pods did not trust the correct CA key and therefore it would roll those pods.

Reply:
I think the general expectation is that every upgrade/downgrade rolls the pods (because of the container images if nothing else). But the downgrade also needs to make sure to maintain the trust to the CAs. So I would assume you need to keep the annotation for that purpose.

Reply:
The downgrade steps I've proposed do roll the pods, but if we removed the annotation the operator would end up rolling the pods twice, rather than just once.